### ISSUE TITLE
--loglevel=silent: command not found when running `make check`

### ISSUE THREAD

**Issue by marxin** [swig/swig#2453]
I must miss a dependency, but can't really see what:

```
LDFLAGS="-fsanitize=undefined,address" CFLAGS="-fsanitize=undefined,address -g" CXXFLAGS="-fsanitize=undefined,address" ./configure &> configure.log.txt
[configure.log.txt](https://github.com/swig/swig/files/10102388/configure.log.txt)
make && make check -j16
...
checking Examples/lua/class
/bin/sh: line 1: --loglevel=silent: command not found
make[3]: *** [../../Makefile:708: javascript_build_cpp] Error 127
make[2]: *** [../example.mk:29: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:215: class.actionexample] Error 2
checking Examples/javascript/constant
checking Examples/csharp/callback
checking Examples/guile/constants
checking errors testcase cpp_bad_global_memberptr
checking guile testcase abstract_inherit
checking errors testcase cpp_class_definition
checking Examples/guile/matrix
checking Examples/lua/constants
/bin/sh: line 1: --loglevel=silent: command not found
make[3]: *** [../../Makefile:708: javascript_build_cpp] Error 127
make[2]: *** [../example.mk:29: build] Error 2
make[2]: Target 'check' not remade because of errors.
make[1]: *** [Makefile:215: constant.actionexample] Error 2
checking Examples/javascript/enum
checking Examples/tcl/constants
checking go testcase go_director_inout (with run test)
```

Thanks for reply.

**Comment by ojwb** [comment]
```
$ git grep loglevel=silent
Examples/Makefile.in:   MAKEFLAGS= $(NODEGYP) --loglevel=silent configure build 1>>/dev/null
Examples/test-suite/javascript/Makefile.in:             MAKEFLAGS= $(COMPILETOOL) $(NODEGYP) --loglevel=silent --directory $* configure build 1>>/dev/null
Examples/test-suite/javascript/Makefile.in:             MAKEFLAGS= $(COMPILETOOL) $(NODEGYP) --loglevel=silent --directory $* configure build 1>>/dev/null
```

So `NODEGYP` is empty.

```
$ git grep NODEGYP configure.ac
configure.ac:    AC_CHECK_PROGS(NODEGYP, node-gyp)
configure.ac:    if test -z "$NODEGYP"; then
configure.ac:AC_SUBST(NODEGYP)
```

So it seems you don't have `node-gyp` installed.

The javascript testsuite should be disabled if required tools or libraries aren't found, and it looks like that ought to happen if `node-gyp` isn't found.  It seems there's a bug here, but I'm not spotting where.

What arguments did you pass to configure (if any)?

What does this report?

```
grep 'JS\|JAVASCRIPT' config.log
```

**Comment by marxin** [comment]
> So it seems you don't have `node-gyp` installed.

Yes, apparently, there's not openSUSE package for it.

> What arguments did you pass to configure (if any)?

No arguments.

```
grep 'JS\|JAVASCRIPT' config.log
ac_cv_prog_NODEJS=node
JSCENABLED='1'
JSCOREDYNAMICLINKING='-ljavascriptcoregtk-4.1 -lgobject-2.0 -lglib-2.0 '
JSCOREINC='-I/usr/include/webkitgtk-4.1 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include '
JSCOREVERSION='2.38.2'
JSINTERPRETERCXX='$(CXX)'
JSINTERPRETERLINKFLAGS='-ldl'
JSV8DYNAMICLINKING=''
JSV8ENABLED=''
JSV8INC=''
LIBOBJS=''
LTLIBOBJS=''
NODEJS=''
SKIP_JAVASCRIPT=''
```

**Comment by marxin** [comment]
Maybe

```
SKIP_JAVASCRIPT=
if test -z "$JAVASCRIPT" || ( test -z "$NODEJS" && test -z "$JSCENABLED" && test -z "$JSV8ENABLED" ) ; then
    SKIP_JAVASCRIPT="1"
fi
```

should also check for `NODEGYP`, what do you think?

**Comment by ojwb** [comment]
That patch would require `node-gyp` to be installed to run tests for JSC or "raw" V8 though, which seems wrong as it's a nodejs-specific tool.

If `node-gyp` is not found, we unset `NODEJS` here:

```  AC_CHECK_PROGS(NODEJS, [node nodejs])

  if test -n "$NODEJS"; then                                                    
    # node-gyp is needed to run the test-suite/examples                         
    AC_CHECK_PROGS(NODEGYP, node-gyp)                                           
    if test -z "$NODEGYP"; then                                                 
      NODEJS=                                                                   
    fi                                                                          
  fi                                                                            
```

So by the time we get to the conditional you suggest patching, if `NODEGYP` was empty then `NODEJS` will also be empty, and `config.log` shows that.

The situation you seem to be in is that JSC has been found, so the bug seems to be that the Javascript tests are trying to run `node-gyp` when `JSENGINE`/`ENGINE` is not set to `node`.

However, all the places where `--loglevel=silent` is used seem to be gated by:

```
ifeq (node,$(JSENGINE))
```

So it looks to me like the problem is how `JSENGINE` gets set. which is this (in two places):

```
ifneq (, $(ENGINE))
        JSENGINE=$(ENGINE)
else
        JSENGINE=node
endif
```

tl;dr: the problem here seems to be that you currently have to run `make check ENGINE=jsc` if you have JSC but not nodejs.

This needs a manual override because you could have both installed and then it's helpful to be able to specify which to test, but really the default shouldn't be hardcoded as `node` as it breaks in a case like yours.  The default should be dynamically chosen so it works (but you can override if you have multiple engines installed and want to test a particular one, as someone maintaining SWIG/JS likely would).

**Comment by ojwb** [comment]
The manual says to use `-node` for nodejs.  The testsuite should really be testing what we tell users to do or there's scope for bugs in the differences.

It seems the only difference in effect between `-node` and `-v8 -DBUILDING_NODE_EXTENSION=1` is that the latter also does:

```
      // V8 API is C++, so output must be C++ compatible even when wrapping C code
      if (!cparse_cplusplus) {
        Swig_cparse_cplusplusout(1);
      }
```

That seems quite a significant difference though.

**Comment by ojwb** [comment]
I've opened #2455 as a first step in sorting this out.
